archive_verify/handlers.py

```diff
@@ -78,40 +79,49 @@ async def status(request):
     q = request.app['redis_q']
     job = q.fetch_job(job_id)
 
-    if job:
-        if job.is_started:
-            payload = {
-                "state": "started",
-                "msg": f"Job {job_id} is currently running."}
-            code = 200
-        elif job.is_finished or job.is_failed:
-            result = job.result
-
-            if result["state"] == "done":
-                payload = {
-                    "state": result["state"],
-                    "msg": f"Job {job_id} has returned with result: {result['msg']}"}
-                code = 200
-            else:
-                payload = {
-                    "state": result["state"],
-                    "msg": f"Job {job_id} has returned with result: {result['msg']}",
-                    "debug": job.exc_info if job.exc_info else result}
-                code = 500
-
-            job.delete()
-        else:
-            payload = {
-                "state": job.get_status(),
-                "msg": f"Job {job_id} is {job.get_status()}"}
-            code = 200
+    if job is None:
+        return web.json_response(
+            {
+                "state": archive_verify.REDIS_STATES["failed"],
+                "msg": f"No such job {job_id} found!"
+            },
+            status=400
+        )
+
+    job_state = archive_verify.REDIS_STATES.get(
+        job.get_status(),
+        "none")
+    code = 200
+    msg = None
+    debug = None
+
+    if job.is_finished or job.is_failed:
+        # this is the dict returned by the worker function
+        result = job.result
+        job_state = result["state"]
+        msg = f"Job {job_id} has returned with result: {result['msg']}"
+
+        if job_state != "done":
+            debug = job.exc_info if job.exc_info else result
+            code = 500
+
+        job.delete()
+    elif job.is_started:
+        msg = f"Job {job_id} is currently running."
     else:
-        payload = {
-            "state": "error",
-            "msg": f"No such job {job_id} found!"}
-        code = 400
-
-    return web.json_response(payload, status=code)
+        msg = f"Job {job_id} is {job.get_status()}"
```

Comment:
I think you should use `job_state` here, i.e. `get_status()`  should only be called once throughout this function, otherwise, there is a risk that it will return different results if the job completes in the middle of the function.

Reply:
Actually I don't think we should use `is_finished`, `is_started` and `is_failed` either, for the same reason. Maybe that's why the state had to be fetched again on L101. Anyways I think the code would be more robust if status was only fetched once.

Reply:
ok, that makes sense. the convenience methods were nice because they abstracted the explicit state values but I agree with your point